src/ws/service/synapse_handler.rs

```diff
@@ -0,0 +1,228 @@
+                        if retry_count == 2 {
+                            return Err(FriendshipsServiceError::InternalServerError.into());
+                        }
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Stores a room event in a Synapse room, and it returns the `EventResponse` containing the event id if the operation was successful
+pub async fn store_room_event_in_synapse_room(
+    token: &str,
+    room_id: &str,
+    room_event: FriendshipEvent,
+    room_message_body: Option<&str>,
+    synapse: &SynapseComponent,
+) -> Result<(), FriendshipsServiceErrorResponse> {
+    let res = synapse
+        .store_room_event(token, room_id, room_event, room_message_body)
+        .await;
+
+    match res {
+        Ok(_) => Ok(()),
+        Err(_) => Err(FriendshipsServiceError::InternalServerError.into()),
+    }
+}
+
+/// Creates a new private room in Synapse and returns the `CreateRoomResponse` if successful.
+/// Returns a `FriendshipsServiceErrorResponse` if there is an error communicating with Synapse.
+///
+/// * `token` - A `&str` representing the auth token.
+/// * `user_ids` - A `Vec<&str>` containing the user ids to invite to the room. There is no need to include the current user id.
+/// * `room_alias_name` -
+/// * `synapse` - A reference to the `SynapseComponent` instance.
+async fn create_private_room_in_synapse(
+    token: &str,
+    user_ids: Vec<&str>,
+    room_alias_name: String,
+    synapse: &SynapseComponent,
+) -> Result<CreateRoomResponse, FriendshipsServiceErrorResponse> {
+    let res = synapse
+        .create_private_room(token, user_ids, &room_alias_name)
+        .await;
+
+    match res {
+        Ok(response) => {
+            let res = CreateRoomResponse {
+                room_id: response.room_id,
+            };
+            Ok(res)
+        }
+        Err(_) => Err(FriendshipsServiceError::InternalServerError.into()),
+    }
+}
+
+async fn get_room_id_for_alias_in_synapse(
+    token: &str,
+    room_alias_name: &str,
+    synapse: &SynapseComponent,
+) -> Result<String, FriendshipsServiceErrorResponse> {
+    let res = synapse.get_room_id_for_alias(token, room_alias_name).await;
+
+    match res {
+        Ok(response) => Ok(response.room_id),
+        Err(_) => Err(FriendshipsServiceError::InternalServerError.into()),
+    }
+}
+
+/// Creates a new Synapse room or returns the existing room id, depending on the `Friendship` and `FriendshipEvent`.
+///
+/// If the `Friendship` exists, returns the `synapse_room_id` in the `Friendship` struct.
+///
+/// If the `Friendship` does not exist and the `FriendshipEvent` is `REQUEST`, checks if a room with the alias exists in Synapse.
+/// If the room exists, returns its id.
+/// If the room does not exist, a new room is created and the new room id is returned.
+///
+/// If the `Friendship` does not exist and the `FriendshipEvent` is not `REQUEST`, an Internal Server Error error is returned.
```

Comment:
Should this be an internal error or a client error due to an invalid request payload?

Reply:
hmm, client error definitely!